redis.go

```diff
@@ -48,83 +48,57 @@ func (hs *hooks) AddHook(hook Hook) {
 func (hs hooks) process(
 	ctx context.Context, cmd Cmder, fn func(context.Context, Cmder) error,
 ) error {
-	ctx, err := hs.beforeProcess(ctx, cmd)
-	if err != nil {
-		cmd.SetErr(err)
-		return err
+	var hookIndex int
+	var retErr error
+
+	for ; hookIndex < len(hs.hooks); hookIndex++ {
+		ctx, retErr = hs.hooks[hookIndex].BeforeProcess(ctx, cmd)
+		if retErr != nil {
+			cmd.SetErr(retErr)
+			break
+		}
 	}
 
-	cmdErr := fn(ctx, cmd)
-
-	if err := hs.afterProcess(ctx, cmd); err != nil {
-		cmd.SetErr(err)
-		return err
+	if retErr == nil {
+		retErr = fn(ctx, cmd)
 	}
 
-	return cmdErr
-}
-
-func (hs hooks) beforeProcess(ctx context.Context, cmd Cmder) (context.Context, error) {
-	for _, h := range hs.hooks {
-		var err error
-		ctx, err = h.BeforeProcess(ctx, cmd)
-		if err != nil {
-			return nil, err
+	for ; hookIndex >= 0; hookIndex-- {
```

Comment:
another issues:
if beforehooks run all success, hookIndex is equals to len(hooks),
afterhooks will be run out of range.

that's why 
using 
```for ; err == nil && hookIndex < len(hs.hooks); hookIndex++ ```
instead of 
```for ;  hookIndex < len(hs.hooks); hookIndex++ ``` is better.

you don't need to check hookIndex is out of range. :)

Reply:
You are right - fixed the code

Reply:
pipelineHooks is also need to fixed :)